src/go/rpk/pkg/config/weak.go

```diff
@@ -0,0 +1,196 @@
+			return fmt.Errorf("cannot parse '%s' as bool: %s", n.Value, err)
+		}
+	default:
+		return fmt.Errorf("type %s not supported as a boolean", n.Tag)
+	}
+}
+
+// weakInt is an intermediary integer type to be used during our transition to
+// strictly typed configuration parameters. This will allow us to support
+// weakly typed parsing:
+//
+//   - strings to int/uint (base implied by prefix)
+//   - bools to int/uint (true = 1, false = 0)
+type weakInt int
+
+func (wi *weakInt) UnmarshalYAML(n *yaml.Node) error {
+	switch n.Tag {
+	case "!!int":
+		ni, err := strconv.Atoi(n.Value)
+		if err != nil {
+			return err
+		}
+		*wi = weakInt(ni)
+		return nil
+	case "!!str":
+		fmt.Println("deprecation warning: type conversion from string to integer will be deprecated")
+		str := n.Value
+		if str == "" {
+			str = "0"
+		}
+		ni, err := strconv.Atoi(str)
+		if err != nil {
+			return fmt.Errorf("cannot parse '%s' as an integer: %s", str, err)
+		}
+		*wi = weakInt(ni)
+		return nil
+	case "!!bool":
+		fmt.Println("deprecation warning: type conversion from boolean to integer will be deprecated")
+		nb, err := strconv.ParseBool(n.Value)
+		if err != nil {
+			return fmt.Errorf("cannot parse '%s' as an integer: %s", n.Value, err)
+		}
+		if nb {
+			*wi = 1
+			return nil
+		}
+		*wi = 0
+		return nil
+	default:
+		return fmt.Errorf("type %s not supported as an integer", n.Tag)
+	}
+}
+
+// weakInt is an intermediary string type to be used during our transition to
+// strictly typed configuration parameters. This will allow us to support
+// weakly typed parsing:
+//
+//   - bools to string (true = "1", false = "0")
+//   - numbers to string (base 10)
+type weakString string
+
+func (ws *weakString) UnmarshalYAML(n *yaml.Node) error {
+	switch n.Tag {
+	case "!!str":
+		*ws = weakString(n.Value)
+		return nil
+	case "!!bool":
+		fmt.Println("deprecation warning: type conversion from boolean to string will be deprecated")
+		nb, err := strconv.ParseBool(n.Value)
+		if err != nil {
+			return fmt.Errorf("cannot parse '%s' as a boolean: %s", n.Value, err)
+		}
+		if nb {
+			*ws = "1"
+			return nil
+		}
+		*ws = "0"
+		return nil
+	case "!!int", "!!float":
+		fmt.Println("deprecation warning: type conversion from numbers to string will be deprecated")
+		*ws = weakString(n.Value)
+		return nil
+	default:
+		return fmt.Errorf("type %s not supported as a string", n.Tag)
+	}
+}
+
+// NamedSocketAddresses is an intermediary one_or_many type to be used
+// during our transition to strictly typed configuration parameters.
+// This type will:
+//   - parse an array of namedSocketAddress
+//   - parse a single namedSocketAddress to an array.
+type NamedSocketAddresses []NamedSocketAddress
+
+func (a *NamedSocketAddresses) UnmarshalYAML(n *yaml.Node) error {
+	var multi []NamedSocketAddress
+	err := n.Decode(&multi)
+	if err == nil {
+		*a = multi
+		return nil
+	}
+
+	fmt.Println("deprecation warning: one_or_many type for namedSocketAddress will be deprecated")
```

Comment:
Maybe `single element named socket addresses in field %q will be deprecated in v23.2.1`